.github/workflows/solana-native.yml

```diff
@@ -11,107 +11,246 @@ on:
+          function build_and_test() {
+            local project=$1
+            local solana_version=$2
+            echo "Building and Testing $project with Solana $solana_version"
+            cd "$project" || return 1
+
+            # Install dependencies
+            if ! pnpm install --frozen-lockfile; then
+              echo "::error::pnpm install failed for $project"
+              echo "$project: pnpm install failed with $solana_version" >> $GITHUB_WORKSPACE/failed_projects.txt
+              cd - > /dev/null
+              return 1
+            fi
+
+            # Build
+            if ! pnpm build; then
+              echo "::error::build failed for $project"
+              echo "$project: build failed with $solana_version" >> $GITHUB_WORKSPACE/failed_projects.txt
+              cd - > /dev/null
+              return 1
+            fi
+
+            # Test
+            if ! pnpm build-and-test; then
+              echo "::error::tests failed for $project"
+              echo "$project: tests failed with $solana_version" >> $GITHUB_WORKSPACE/failed_projects.txt
+              cd - > /dev/null
+              return 1
+            fi
+
+            echo "Build and tests succeeded for $project with $solana_version version."
+            cd - > /dev/null
+            return 0
+          }
+
+          function process_projects() {
+            local solana_version=$1
+
+            readarray -t all_projects < <(echo '${{ needs.changes.outputs.changed_projects }}' | jq -r '.[]?')
+            start_index=$(( ${{ matrix.index }} * ${{ env.MIN_PROJECTS_PER_JOB }} ))
+            end_index=$(( start_index + ${{ env.MIN_PROJECTS_PER_JOB }} ))
+            end_index=$(( end_index > ${{ needs.changes.outputs.total_projects }} ? ${{ needs.changes.outputs.total_projects }} : end_index ))
+
+            echo "Projects to build and test in this job"
+            for i in $(seq $start_index $(( end_index - 1 ))); do
+              echo "${all_projects[$i]}"
+            done
+
+            failed=false
+            for i in $(seq $start_index $(( end_index - 1 ))); do
+              echo "::group::Building and testing ${all_projects[$i]}"
+              if ! build_and_test "${all_projects[$i]}" "$solana_version"; then
+                failed=true
+              fi
+              echo "::endgroup::"
+            done
+
+            return $([ "$failed" = true ] && echo 1 || echo 0)
+          }
+          EOF
+
+          # Make the script executable
+          chmod +x build_and_test.sh
+
+          # Install pnpm
           npm install --global pnpm
-      - name: Test solana native programs
+      - name: Setup Solana stable
+        uses: heyAyushh/setup-solana@v5.5
+        with:
+          solana-cli-version: stable
+      - name: Build and Test with Stable
+        run: |
+          source build_and_test.sh
+          solana -V
+          rustc -V
+          process_projects "stable"
+      - name: Setup Solana 1.18.17
```

Comment:
Why not use the solana version of the project?

Reply:
You could use verified builds for this maybe. It has docker images for all solana versions

Reply:
This 1.18.17 version is the minimum supported solana version for this repository, stable is to track upcoming bugs or failures.

Reply:
Verified builds should be another PR I think.

Reply:
Yeah for sure. Just an idea